allow service NodePort to be updated to 0 in case AllocateLoadBalance…

armstrongli opened the pull request:
…rNodePorts=false

the original logic always guarantee the NodePort's value if it was there. the NodePort should be allowed to set 0 if the Service has LB type with AllocateLoadBalancerNodePorts=false


/kind bug


#### What this PR does / why we need it:

`AllocateLoadBalancerNodePorts: false` means the LoadBalancer type service doesn't require NodePorts in the ports. the node ports are created on Service creation while landing to etcd storage. it is also allowed to update the field on the fly. the NodePorts are not recycled on the fly. neither the removal on updates because storage keeps trying put back the old value if the old value is still valid.

it is needed to allow manual remove the NodePorts by service owner while AllocateLoadBalancerNodePorts=false.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

HirazawaUi commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -625,6 +625,12 @@ func patchAllocatedValues(after After, before Before) {
+				// skip setting NodePort in case AllocateLoadBalancerNodePorts is false
+				if newSvc.Spec.Type == api.ServiceTypeLoadBalancer &&
+					newSvc.Spec.AllocateLoadBalancerNodePorts != nil &&
+					*newSvc.Spec.AllocateLoadBalancerNodePorts == false {
+					continue
+				}
```
```suggestion
					!*newSvc.Spec.AllocateLoadBalancerNodePorts {
					continue
				}
```

armstrongli replied:
we can't use the *AllocateLoadBalancerNodePorts directly as the value can be null and bring nil pointer panics. especially while upgrading control plane without updating the service specs.

HirazawaUi replied:
I mean we just need to use `!*newSvc.Spec.AllocateLoadBalancerNodePorts` instead of `*newSvc.Spec.AllocateLoadBalancerNodePorts == false`

armstrongli replied:
updated.

HirazawaUi commented:
/sig network

aojea commented:
I may not understand the KEP says

https://github.com/kubernetes/enhancements/tree/b103a6b0992439f996be4314caf3bf7b75652366/keps/sig-network/1864-disable-lb-node-ports#design-details

> API changes to Service:
> Add a new field spec.allocateLoadBalancerNodePorts: true|false.
> allocateLoadBalancerNodePorts defaults to true, preserving existing behavior for Service Type=LoadBalancer.
> On create, if allocateLoadBalancerNodePorts: false, don't allocate node ports.
> On update, if allocateLoadBalancerNodePorts: false don't allocate new node ports but do not deallocate existing node ports if set.
> On update, set nodePort: 0 (by removing the field) to deallocate a node port, a new port will not be re-allocated if allocateLoadBalancerNodePorts: false.
> On delete, node ports are deallocated regardless of allocateLoadBalancerNodePorts.


 can you add an example on how to reproduce the problem? 

anyway, if there is need for a change, it will need an unit test?

armstrongli commented:
> can you add an example on how to reproduce the problem?

```bash
:kind-capi-2:default:~# k get svc rest-1 -o yaml | yq .spec
allocateLoadBalancerNodePorts: false
clusterIP: 10.109.145.199
clusterIPs:
  - 10.109.145.199
externalTrafficPolicy: Cluster
internalTrafficPolicy: Cluster
ipFamilies:
  - IPv4
ipFamilyPolicy: SingleStack
ports:
  - name: http
    nodePort: 32428
    port: 8080
    protocol: TCP
    targetPort: 8080
selector:
  app: rest
sessionAffinity: None
type: LoadBalancer
:kind-capi-2:default:~# k patch svc rest-1 -p '{"spec":{"$setElementOrder/ports":[{"port":8080}],"ports":[{"nodePort":null,"port":8080}]}}'
service/rest-1 patched (no change)
:kind-capi-2:default:~# k get svc rest-1 -o yaml | yq .spec
allocateLoadBalancerNodePorts: false
clusterIP: 10.109.145.199
clusterIPs:
  - 10.109.145.199
externalTrafficPolicy: Cluster
internalTrafficPolicy: Cluster
ipFamilies:
  - IPv4
ipFamilyPolicy: SingleStack
ports:
  - name: http
    nodePort: 32428
    port: 8080
    protocol: TCP
    targetPort: 8080
selector:
  app: rest
sessionAffinity: None
type: LoadBalancer
```

> anyway, if there is need for a change, it will need an unit test?

ack. i'll add UT in my daytime.

aojea commented:
>  k patch svc rest-1 -p '{"spec":{"$setElementOrder/ports":[{"port":8080}],"ports":[{"nodePort":null,"port":8080}]}}'
service/rest-1 patched (no change)

the nodePort field is an int32, a patch with the default value will not set it to zero right?

aojea commented:
that is my point , the patch will interpret that value as "no change needed" AFAIK, is not way it can differentiate it from "set this to 0"

armstrongli commented:
> that is my point , the patch will interpret that value as "no change needed" AFAIK, is not way it can differentiate it from "set this to 0"

setting the value to be `null` means removing the field from the port defintion. the unmarshal in the backend make the int value as default - 0 . finally the code goes into the logic that the new svc's nodePort value is 0 . the nodePort==0 part brings the old port value and check the usability and reuse it before landing the object into storage. 

the returned payload from apiserver to kubelet has exactly same specs and kubectl prints the no changes.

armstrongli commented:
@aojea positive/negative UTs added

aojea commented:
> @aojea positive/negative UTs added

/assign @aojea

> setting the value to be null means removing the field from the port defintion.

let me check this in more detail, I think that is not how is this happening,

HirazawaUi commented:
According to the [proposal](https://github.com/kubernetes/enhancements/tree/b103a6b0992439f996be4314caf3bf7b75652366/keps/sig-network/1864-disable-lb-node-ports#proposal), the current behavior is not as expected, even if we explicitly remove the nodePort field using kubectl edit, the nodePort will still exist.

And some [original discussion](https://github.com/kubernetes/enhancements/pull/1514#issuecomment-644787951) FYI.

armstrongli commented:
add issue: https://github.com/kubernetes/kubernetes/issues/122381

aojea commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -625,6 +625,12 @@ func patchAllocatedValues(after After, before Before) {
 		for i := range newSvc.Spec.Ports {
 			p := &newSvc.Spec.Ports[i]
 			if p.NodePort == 0 {
+				// skip setting NodePort in case AllocateLoadBalancerNodePorts is false
+				if newSvc.Spec.Type == api.ServiceTypeLoadBalancer &&
+					newSvc.Spec.AllocateLoadBalancerNodePorts != nil &&
+					!*newSvc.Spec.AllocateLoadBalancerNodePorts {
+					continue
+				}
```
this will modify the behavior, a patch can not deallocate the nodePorts

aojea replied:
you have to remove this

armstrongli replied:
updated

aojea commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -648,9 +654,19 @@ func needsClusterIP(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeNodePort {
 		return true
 	}
+	if svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		if svc.Spec.AllocateLoadBalancerNodePorts == nil || *svc.Spec.AllocateLoadBalancerNodePorts {
+			return true
+		}
+		for _, item := range svc.Spec.Ports {
+			if item.NodePort > 0 {
+				return true
+			}
+		}
+	}
```
I'm not sure this heuristic here is correct, there are at least other function in strategy that replicates this logic that we need to check too

armstrongli replied:
i'll take a look the strategy part, and add corresponding integration tests, e2e tests.

aojea replied:
to be clear, I don't think that the existence of NodePorts should be part of this function,

armstrongli replied:
right. after digging more, this part is not required. removed.

aojea commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -648,9 +654,19 @@ func needsClusterIP(svc *api.Service) bool {
 		return true
 	}
+	if svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		if svc.Spec.AllocateLoadBalancerNodePorts == nil || *svc.Spec.AllocateLoadBalancerNodePorts {
+			return true
+		}
```
at this point it was already defaulted IIRC https://github.com/kubernetes/kubernetes/blob/6aac45ff1e99068e834ba3b93b673530cf62c007/pkg/apis/core/v1/defaults.go#L139-L143

armstrongli replied:
we need the nil check because the field was imported later and the legacy objects don't have the filed. and the unmarshal gives it a `null` and bring apiserver nil pointer panics.

armstrongli committed:
allow service NodePort to be updated to 0 in case AllocateLoadBalancerNodePorts=false

the original logic always guarantee the NodePort's value if it was there. the NodePort should be allowed to set 0 if the Service has LB type with AllocateLoadBalancerNodePorts=false

armstrongli commented:
@aojea  code updated. please help review

aojea commented:
/ok-to-test

aojea commented on test/integration/service/loadbalancer_test.go:
```diff
@@ -178,6 +178,115 @@ func Test_ServiceLoadBalancerEnableThenDisableAllocatedNodePorts(t *testing.T) {
+
+	service.Spec.AllocateLoadBalancerNodePorts = utilpointer.Bool(false)
+	service.Spec.Ports[0].NodePort = 0
+	service, err = client.CoreV1().Services(ns.Name).Update(context.TODO(), service, metav1.UpdateOptions{})
```
we need to add another test case using Patch instead of Update

armstrongli replied:
case added. please review

armstrongli committed:
add patch integration test to disable nodeport for LB type service

armstrongli commented:
/retest pull-kubernetes-e2e-gce

armstrongli commented:
/test pull-kubernetes-e2e-gce

aojea commented on test/integration/service/loadbalancer_test.go:
```diff
@@ -178,6 +181,188 @@ func Test_ServiceLoadBalancerEnableThenDisableAllocatedNodePorts(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Error marshalling test service: %v", err)
+	}
+	patch, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, corev1.Service{})
```
is this equal to  your previous example `k patch svc rest-1 -p '{"spec":{"$setElementOrder/ports":[{"port":8080}],"ports":[{"nodePort":0,"port":8080}]}}' `

We have different patch mechanism, should we add a table case with the different types of patches?

I also would be interested in create a patch that ONLY changes the `AllocateLoadBalancerNodePorts` field to false and validates that no NodePort is removed

armstrongli replied:
yes. the patch is same. the NodePort field is labelled with `omitempty`. then the patch `nodePort":0` is equal to `nodePort": null`. patch happens in the serving REST part. the change we do is after the REST. the object is convered from external version object to internal version object. the final difference is same to have a `spec.ports[*].nodePort=0`.

aojea replied:
thanks for clarifying

aojea commented:
/assign @liggitt @thockin 

Overall it looks good, however, the patching on Services always give me headaches, I will try to do more local experimentation, but the tests added gives some confidence on the change

armstrongli commented:
@liggitt @thockin , please help review this PR. it is about a fix for updating NodePort to 0 when allocate LB node port is false.

thockin commented:
Thanks.  Given the imminent KEP freeze, I am focusing on KEPs right now.  I will circle back to this ASAP.

thockin commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -648,7 +648,11 @@ func needsClusterIP(svc *api.Service) bool {
+		return true
+	}
+	if svc.Spec.Type == api.ServiceTypeLoadBalancer &&
+		(svc.Spec.AllocateLoadBalancerNodePorts == nil || *svc.Spec.AllocateLoadBalancerNodePorts) {
```
You should not need to check nil - API guarantees it won't be nil.

thockin reviewed (commented):
As Antonio pointed out, the KEP says:

> * On update, if allocateLoadBalancerNodePorts: false don't allocate new node ports but do not deallocate existing node ports if set.
> * On update, set nodePort: 0 (by removing the field) to deallocate a node port, a new port will not be re-allocated if allocateLoadBalancerNodePorts: false.

The NodePort field is defined as `int`, not a `*int`.  That means that if you say nothing, the apiserver will observe that field as "set to 0" on a create or update (GET / PUT).  We don't (can't) know if you meant to set it to 0 or if you just didn't say anything.

Patch and apply operations are OK.

In other words, there's no way both of the above goals from the KEP can be met on a `kubectl replace`.  The question then, is whether we think it is more likely that the user meant to retain the nodeport or to wipe it away.

They *did* set `allocateLoadBalancerNodePorts` to false, which is pretty explicit, I think.

@aojea - agree?

thockin commented:
/approve

aojea commented:
/lgtm
